### PR TITLE
Allow dtype-convertible values in reductions

### PR DESCRIPTION
**Pull request by alexfikl**
Allows passing `np.float64` instead of a dtype to these functions. Most of the kernels (e.g. `get_sum_kernel`) assume they're getting a dtype, e.g.

https://github.com/inducer/pyopencl/blob/3abb137cb4dfe78637fd670d0657928d3cce2460/pyopencl/reduction.py#L529

### CODE REVIEW

**Comment by inducer** [comment]
Thanks, looks good. Add a test?

**Comment by alexfikl** [comment]
> Thanks, looks good. Add a test?

@inducer Added a little test for `float` values (and a lot of markers!).

File: test/test_array.py [hunk 1]
```diff
@@ -1788,6 +2060,39 @@ def test_empty_reductions_vs_numpy(ctx_factory, reduction, supports_initial):
...
+
+    result = cl.array.vdot(ary, ary, dtype=np.float32)
+    assert result.dtype == np.float32, result.dtype
+    result = cl.array.vdot(ary, ary, dtype=np.float64)
+    assert result.dtype == np.float64, result.dtype
+
+    result = cl.array.cumsum(ary, output_dtype=np.float32)
+    assert result.dtype == np.float32, result.dtype
+    result = cl.array.cumsum(ary, output_dtype=np.float64)
+    assert result.dtype == np.float64, result.dtype
```
Comment (inducer) [inline test/test_array.py]: Convert to two loops?

File: test/test_array.py [hunk 2]
```diff
@@ -1788,6 +2060,39 @@ def test_empty_reductions_vs_numpy(ctx_factory, reduction, supports_initial):
 
     # }}}
 
+# }}}
+
+
+# {{{ test_reductions_dtype
+
+def test_reductions_dtype(ctx_factory):
```
Comment (inducer) [inline test/test_array.py]: Technically, `cumsum` isn't a reduction. :)

**Review by inducer — APPROVED** [review]
Thanks! And: Sorry. Found two nitpicks. If you can't be bothered, let me know, and I'll merge anyhow.

File: test/test_array.py (hunk 2 above)
Comment (alexfikl) [reply test/test_array.py]: Renamed the test function in https://github.com/inducer/pyopencl/compare/3ba38d1f48603a1ab32250a5fe75ef00999b35da..61347c18df669418ac9d0c4becfdb8afdd6a776c.

File: test/test_array.py (hunk 1 above)
Comment (alexfikl) [reply test/test_array.py]: Done in https://github.com/inducer/pyopencl/compare/3ba38d1f48603a1ab32250a5fe75ef00999b35da..61347c18df669418ac9d0c4becfdb8afdd6a776c.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**